AsyncDisplayKit/ASImageNode+AnimatedImage.mm

```diff
@@ -0,0 +1,226 @@
+//
+//  ASImageNode+AnimatedImage.m
+//  AsyncDisplayKit
+//
+//  Created by Garrett Moon on 3/22/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "ASImageNode+AnimatedImage.h"
+
+#import "ASAssert.h"
+#import "ASImageProtocols.h"
+#import "ASDisplayNode+Subclasses.h"
+#import "ASDisplayNodeExtras.h"
+#import "ASEqualityHelpers.h"
+#import "ASDisplayNode+FrameworkPrivate.h"
+#import "ASImageNode+AnimatedImagePrivate.h"
+#import "ASInternalHelpers.h"
+#import "ASWeakProxy.h"
+
+@implementation ASImageNode (AnimatedImage)
+
+#pragma mark - GIF support
+
+- (void)setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  if (!ASObjectIsEqual(_animatedImage, animatedImage)) {
+    _animatedImage = animatedImage;
+  }
+  if (animatedImage != nil) {
+    __weak ASImageNode *weakSelf = self;
+    if ([animatedImage respondsToSelector:@selector(setCoverImageReadyCallback:)]) {
+      animatedImage.coverImageReadyCallback = ^(UIImage *coverImage) {
+        [weakSelf coverImageCompleted:coverImage];
+      };
+    }
+    
+    animatedImage.playbackReadyCallback = ^{
+      [weakSelf animatedImageFileReady];
+    };
+  }
+}
+
+- (id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  return _animatedImage;
+}
+
+- (void)setAnimatedImagePaused:(BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
+  _animatedImagePaused = animatedImagePaused;
+  ASPerformBlockOnMainThread(^{
+    if (animatedImagePaused) {
+      [self stopAnimating];
+    } else {
+      [self startAnimating];
+    }
+  });
+}
+
+- (BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
+  return _animatedImagePaused;
+}
+
+- (void)coverImageCompleted:(UIImage *)coverImage
+{
+  BOOL setCoverImage = YES;
+  {
+    ASDN::MutexLocker l(_displayLinkLock);
+    if (_displayLink != nil && _displayLink.paused == NO) {
+      setCoverImage = NO;
+    }
+  }
+  
+  if (setCoverImage) {
+    self.image = coverImage;
+  }
+}
+
+- (void)animatedImageFileReady
+{
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [self startAnimating];
+  });
+}
+
+- (void)startAnimating
+{
+  ASDisplayNodeAssertMainThread();
+  if (ASInterfaceStateIncludesVisible(self.interfaceState) == NO) {
+    return;
+  }
+  
+  if (self.animatedImagePaused) {
+    return;
+  }
+  
+  if (self.animatedImage.playbackReady == NO) {
+    return;
+  }
+  
+#if ASAnimatedImageDebug
+  NSLog(@"starting animation: %p", self);
+#endif
+  ASDN::MutexLocker l(_displayLinkLock);
+  if (_displayLink == nil) {
+    _playHead = 0;
+    _displayLink = [CADisplayLink displayLinkWithTarget:[ASWeakProxy weakProxyWithTarget:self] selector:@selector(displayLinkFired:)];
```

Comment:
Why does this require a weak proxy? I think this requires a comment at this location, but ideally we could eliminate it with some other construct.

If we keep the new proxy class, there should also be at least a brief description at the beginning of the class of an example use case. If we can't think of another use case to document it with, I'm not certain we should define it as a generalized primitive.

 Although it is true that it would be harder to find if it is eventually needed by another developer.